lambda/translquote.ml

```diff
@@ -0,0 +1,3144 @@
+
+let with_new_idents_types_constr =
+  List.iter (fun id -> with_type_constr id (Lvar id))
+
+let with_new_idents_modules = List.iter (fun id -> with_module id (Lvar id))
+
+let without_value ident = Hashtbl.remove vars_env.env_vals ident
+
+let without_type_constr ident = Hashtbl.remove vars_env.env_tys ident
+
+let without_module ident = Hashtbl.remove vars_env.env_mod ident
+
+let without_idents_values = List.iter without_value
+
+let without_idents_types_constr = List.iter without_type_constr
+
+let without_idents_modules = List.iter without_module
+
+let with_poly_type name =
+  let id = Ident.create_local name in
+  Hashtbl.add vars_env.env_poly name (id, Var.Type_var.mk (Lvar id))
+
+let with_new_idents_poly = List.iter with_poly_type
+
+let without_poly_type name = Hashtbl.remove vars_env.env_poly name
+
+let without_idents_poly = List.iter without_poly_type
+
+let ident_for_poly_name name =
+  let ident, _ = Hashtbl.find vars_env.env_poly name in
+  ident
+
+(* Compiling quoted expressions to lambda *)
+
+let use modname field = combinator modname field
+
+let apply modname field loc args =
+  lazy
+    (let comb = use modname field in
+     Lambda.Lapply
+       { ap_func = Lazy.force comb;
+         ap_args = args;
+         ap_probe = None;
+         ap_loc = of_location ~scopes:empty_scopes loc;
+         ap_result_layout =
+           Pvalue { raw_kind = Pgenval; nullable = Non_nullable };
+         ap_region_close = Rc_normal;
+         ap_mode = alloc_heap;
+         ap_tailcall = Default_tailcall;
+         ap_inlined = Default_inlined;
+         ap_specialised = Default_specialise
+       })
+
+let apply1 modname field loc arg1 = apply modname field loc [arg1]
+
+let apply2 modname field loc arg1 arg2 = apply modname field loc [arg1; arg2]
+
+let apply3 modname field loc arg1 arg2 arg3 =
+  apply modname field loc [arg1; arg2; arg3]
+
+let apply4 modname field loc arg1 arg2 arg3 arg4 =
+  apply modname field loc [arg1; arg2; arg3; arg4]
+
+let apply5 modname field loc arg1 arg2 arg3 arg4 arg5 =
+  apply modname field loc [arg1; arg2; arg3; arg4; arg5]
+
+let apply6 modname field loc arg1 arg2 arg3 arg4 arg5 arg6 =
+  apply modname field loc [arg1; arg2; arg3; arg4; arg5; arg6]
+
+module Loc : sig
+  type s
+
+  type t' = s lazy_t
+
+  type t = s lam
+
+  val wrap : t' -> t
+
+  val unknown : t'
+
+  val known : Location.t -> string -> int -> int -> int -> int -> t'
+end = struct
+  type s = lambda
+
+  type t' = s lazy_t
+
+  type t = s lam
+
+  let wrap = inject_force
+
+  let unknown = use "Loc" "unknown"
+
+  let known loc file start_line start_col end_line end_col =
+    apply5 "Loc" "known" loc
+      (Lconst (Const_base (Const_string (file, loc, None))))
+      (Lconst (Const_base (Const_int start_line)))
+      (Lconst (Const_base (Const_int start_col)))
+      (Lconst (Const_base (Const_int end_line)))
+      (Lconst (Const_base (Const_int end_col)))
+end
+
+module Name : sig
```

Comment:
It looks generally true that these modules are more complicated than they need to be `Name.mk` is always immediately followed by `Name.wrap`.

Reply:
This has everything to do with the "knot" that makes the `CamlinternalQuote` and `Translquote` signatures basically the same.